### PR TITLE
Ember: Stop using acceptance tests for everything

### PR DESCRIPTION
**Pull request by martndemus**
Reserving the topic, will write words very soon.

### CODE REVIEW

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+- Testing components with acceptance tests is wasting a lot of time by booting a full app and navigating to the page your component is on, etc...
+- Nobody likes to wait 5 to 10 minutes on a test suite to pass.
+- Introduce the integration test, you can test anything template related. 
+- They are also the better alternative to test things that are hard to test with unit tests. (e.g. models with deps on relationships).
+- Sadly it has completely different mechanics respective to acceptance tests, this will be fixed 'soon'.
+- Resources for learning how to integration test stuff properly.
+- .....
+- Profit!!!
+
+Should I show some examples?
```
Comment (brzpegasus) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: Some concrete examples of what's suitable to be tested in an acceptance test vs integration test might be sufficient.

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+social: true
+author: Marten Schilstra
+twitter: "martndemus"
+github: martndemus
+summary: "A brief summary of your post"
+published: true
+tags: tags, separating each, with commas
+---
+
+- In Ember it is really easy to test with acceptance tests, you only have to recreate what you do in the browser
```
Comment (bcardarella) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: Are these meant to be bullet lists in the final post or just placeholders for you to expand upon later?

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
+---
+layout: post
+title: "Stop using acceptance tests for everything"
+social: true
+author: Marten Schilstra
+twitter: "martndemus"
+github: martndemus
+summary: "A brief summary of your post"
```
Comment (rsocci) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: don't forget to add a real summary here :)

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+published: true
+tags: tags, separating each, with commas
```
Comment (rsocci) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: and tags

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md (hunk 2 above)
Comment (martndemus) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: These are here for my story line, they will be gone in the final result.

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+- .....
+- Profit!!!
+
+Should I show some examples?
+
+# Prefer Integration Tests over Acceptance Tests
+
+It is suprisingly easy to test your whole Ember.JS application using acceptance tests only. All you have to do is to reproduce how you would normally navigate through the application using a [stellar set of helpers](https://guides.emberjs.com/v2.4.0/testing/acceptance/). Don't wan't to test your API? [Add ember-cli-mirage](http://www.ember-cli-mirage.com)!
+
+You should realize though, that each acceptance test boots up the whole ember application and starts you at the `application.index` route. You then have to navigate to the right page and then test the thing you wanted to do. This is fine when you want to test a complex user-flow through an app, but a big waste of time and cpu cycles to verify if a form submit button is unclickable when the data is invalid.
```
Comment (rsocci) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: capitalize `Ember`

Comment (rsocci) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: uppercase `CPU`

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+If you create acceptance tests for every edge case and every regression, then it will be likely that your test suite will quickly grow to being five, ten or more minutes long. That is a problem, slow test suites will get run less often, test failures won't be caught quickly. I have caught myself relying on the CI server to run the tests for me, while doing other stuff, because I didn't want to wait for just five minutes.
```
Comment (rsocci) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: quickly grow to *be* five...

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+You should realize though, that each acceptance test boots up the whole Ember.JS application and starts you at the `application.index` route. You then have to navigate to the right page and then test the thing you wanted to do. This is fine when you want to test a complex user-flow through an app, but a big waste of time and CPU cycles to verify if a form submit button is unclickable when the data is invalid.
+
+If you create acceptance tests for every edge case and every regression, then it will be likely that your test suite will quickly grow to be five, ten or more minutes long. That is a problem, slow test suites will get run less often, test failures won't be caught quickly. I have caught myself relying on the CI server to run the tests for me, while doing other stuff, because I didn't want to wait for just five minutes.
```
Comment (rsocci) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: Wording suggestion: That is a problem since slow test suites are likely to be run less often by the developer, and test failures won't be caught quickly.

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+Having said all that, I think we should write more integration tests and less acceptance tests. Integration tests in most cases run much faster, because you don't have to boot a whole app, or navigate to the page where you component is on.
```
Comment (rsocci) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: where *your* component is *being consumed*

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+Integration tests are also more flexible then most people think. It's not only for testing components, you can also test [template helpers](https://github.com/DockYard/ember-composable-helpers/blob/master/tests/integration/helpers/map-by-test.js). Actually you can test anything with an [integration test](https://github.com/switchfly/ember-test-helpers/blob/master/lib/ember-test-helpers/test-module.js#L26). For example a model that has logic depending on it's relationships is a good example where you can use a non-component integration test.
```
Comment (rsocci) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: `then` -> `than`

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
```
Comment (rsocci) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: *They're* not only for testing, but can also be used to test...

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+Having said all that, I think we should write more integration tests and less acceptance tests. Integration tests in most cases run much faster, because you don't have to boot a whole app, or navigate to the page where your component is being used.
+
+Integration tests are also more flexible then most people think. They're not only for testing components, you can also test [template helpers](https://github.com/DockYard/ember-composable-helpers/blob/master/tests/integration/helpers/map-by-test.js). Actually you can test anything with an [integration test](https://github.com/switchfly/ember-test-helpers/blob/master/lib/ember-test-helpers/test-module.js#L26). For example a model that has logic depending on it's relationships is a good example where you can use a non-component integration test.
```
Comment (rsocci) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: depending on `its`

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+Integration tests are also more flexible than most people think. They're not only for testing components, you can also test [template helpers](https://github.com/DockYard/ember-composable-helpers/blob/master/tests/integration/helpers/map-by-test.js). Actually you can test anything with an [integration test](https://github.com/switchfly/ember-test-helpers/blob/master/lib/ember-test-helpers/test-module.js#L26). For example a model that has logic depending on it's relationships is a good example where you can use a non-component integration test.
+
+There is a small downside to integration tests: the api for writing component integration tests is completely different than the api for writing acceptance tests, so you will have to learn another totally different style of writing tests while maintaining the knowledge of how to write acceptance tests. Luckily there is a [motion to unify](https://github.com/emberjs/rfcs/pull/119) the api for integration and acceptance tests.
```
Comment (rsocci) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: capitalize `API`

File: source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+Integration tests are also more flexible than most people think. They're not only for testing components, you can also test [template helpers](https://github.com/DockYard/ember-composable-helpers/blob/master/tests/integration/helpers/map-by-test.js). Actually you can test anything with an [integration test](https://github.com/switchfly/ember-test-helpers/blob/master/lib/ember-test-helpers/test-module.js#L26). For example a model that has logic depending on its relationships is a good example where you can use a non-component integration test.
+
+There is a small downside to integration tests: the api for writing component integration tests is completely different than the api for writing acceptance tests, so you will have to learn another totally different style of writing tests while maintaining the knowledge of how to write acceptance tests. Luckily there is a [motion to unify](https://github.com/emberjs/rfcs/pull/119) the api for integration and acceptance tests.
+
```
Comment (rsocci) [inline source/posts/2016-03-25-stop-using-acceptance-tests-for-everything.md]: kill extra line :fire:

**Comment by rsocci** [comment]
I like the post so far! :+1:

File: source/posts/2016-03-28-stop-using-acceptance-tests-for-everything.md [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+social: true
+author: Marten Schilstra
+twitter: "martndemus"
+github: martndemus
+summary: "Integration testing should be an integral part of your test suite."
+published: true
+tags: ember, javascript, testing
+---
+
+It is suprisingly easy to test your whole Ember.JS application using acceptance tests only. All you have to do is to reproduce how you would normally navigate through the application using a [stellar set of helpers](https://guides.emberjs.com/v2.4.0/testing/acceptance/). Don't wan't to test your API? [Add ember-cli-mirage](http://www.ember-cli-mirage.com)!
```
Comment (poteto) [inline source/posts/2016-03-28-stop-using-acceptance-tests-for-everything.md]: `s/Don't wan't to test your API?/Don't want to test your API?`

File: source/posts/2016-03-28-stop-using-acceptance-tests-for-everything.md [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+Using component integration tests to test components should also improve your component design. Test driving individual components will make you think more about their external api, prevents building a [LMAO](http://slides.com/miguelcamba/composable-components#/8) and promotes [loose coupling](https://en.wikipedia.org/wiki/Loose_coupling).
+
+There is, however, a small downside to integration tests: the API for writing component integration tests is completely different than the api for writing acceptance tests, so you will have to learn another totally different style of writing tests while maintaining the knowledge of how to write acceptance tests. Luckily there is a [motion to unify](https://github.com/emberjs/rfcs/pull/119) the API for integration and acceptance tests.
```
Comment (poteto) [inline source/posts/2016-03-28-stop-using-acceptance-tests-for-everything.md]: `s/than the api for/than the API for`

File: source/posts/2016-03-28-stop-using-acceptance-tests-for-everything.md (hunk 14 above)
Comment (poteto) [inline source/posts/2016-03-28-stop-using-acceptance-tests-for-everything.md]: Maybe just use `Ember` instead of `Ember.JS`?
